pkg/daemon/daemon.go

```diff
@@ -424,6 +440,75 @@ func (dn *Daemon) getPendingConfig() (string, error) {
 	return p.PendingConfig, nil
 }
 
+// setupFileWatcher sets up a watch on the SSH taint file via a goroutine,
+// and applies a NoSchedule taint to the node when it detects a write to
+// the file (performed by sshd)
+func (dn *Daemon) setupFileWatcher(watcher *fsnotify.Watcher) error {
+	go func() {
+		for {
+			select {
+			// watch for events on the file
+			case event := <-watcher.Events:
+				glog.V(2).Infof("Taint file watcher received event: %#v\n", event)
+				fileInfo, err := os.Stat(pathSSHTaint)
+				if err != nil {
+					glog.Error("Error during taint: cannot get fileInfo in Stat function.")
+				}
+
+				// Check if the ssh script wrote to the file
+				if fileInfo.Size() != 0 {
+					if err := dn.applySSHTaint(); err != nil {
+						glog.Errorf("Error during taint: cannot apply taint due to: %v", err)
+					}
+					// Now that the taint has been applied, we truncate the file to watch for future ssh's
+					_, err = os.Create(pathSSHTaint)
+					if err != nil {
+						glog.Infof("Error during taint: cannot truncate the file.")
+					}
+				}
+				// Re-watch the file for future SSH access
+				watcher.Add(pathSSHTaint)
+
+			case err := <-watcher.Errors:
+				glog.Errorf("Error watching runtime file: %v", err)
+			}
+		}
+	}()
+
+	// first check if there was a taint from before the reboot that we did not catch
+	if fileInfo, err := os.Stat(pathSSHTaint); !os.IsNotExist(err) && fileInfo.Size() != 0 {
+		if err = dn.applySSHTaint(); err != nil {
+			glog.Errorf("Error during taint: cannot apply taint due to: %v", err)
+		}
+
+	}
+
+	// Now create (truncate) the file to set up the watcher
+	if _, err := os.Create(pathSSHTaint); err != nil {
+		return err
+	}
+
+	if err := watcher.Add(pathSSHTaint); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// applySSHTaint calls nodewriter to apply the specified ssh taint
+func (dn *Daemon) applySSHTaint() error {
+	glog.Infof("Detected ssh! The node is being tainted with: %v=%v:%v",
+	MachineConfigDaemonSSHTaintKey, MachineConfigDaemonSSHTaintValue, corev1.TaintEffectNoSchedule)
+
+	taint := &corev1.Taint {
+		Key: MachineConfigDaemonSSHTaintKey,
+		Value: MachineConfigDaemonSSHTaintValue,
+		Effect: corev1.TaintEffectNoSchedule,
+	}
+	return dn.nodeWriter.SetTaint(dn.kubeClient.CoreV1().Nodes(), dn.name, taint)
```

Comment:
Optional: would be nice to also register an event here when tainting to give it higher visibility.